sources/ghostdog/src/infiniband.rs

```diff
@@ -0,0 +1,197 @@
+        .join(SW_MNG_CLASS)
+        .join(device)
+        .join("ports");
+
+    let port_directories = fs::read_dir(ports_path).context(error::InfinibandSysDevicesSnafu)?;
+    for port in port_directories {
+        let mut capability_mask: String = "".to_string();
+        let mut first_guid: String = "".to_string();
+
+        let port = port.context(error::InfinibandSysDevicesSnafu)?;
+        let capability_mask_path = port.path().join("cap_mask");
+        if capability_mask_path.exists() {
+            capability_mask = std::fs::read_to_string(capability_mask_path.as_path()).context(
+                error::ReadFileSnafu {
+                    path: capability_mask_path.as_path(),
+                },
+            )?;
+        }
+        let first_guid_path = port.path().join("gids").join("0");
+        if first_guid_path.exists() {
+            first_guid = std::fs::read_to_string(first_guid_path.as_path()).context(
+                error::ReadFileSnafu {
+                    path: first_guid_path.as_path(),
+                },
+            )?;
+        }
+        // If a capability mask and port guid have been read, create the Infiniband Port
+        if !capability_mask.is_empty() && !first_guid.is_empty() {
+            found_ports.push(InfinibandPort {
+                capability_mask: CapabilityMask::from_str(capability_mask.as_str())?,
+                port_guid: Guid::from_str(first_guid.as_str())?,
+            });
+        }
+    }
+    Ok(found_ports)
+}
+
+/// Guid struct to help with custom FromStr and ToString
+#[derive(Clone, Debug)]
+pub(crate) struct Guid(String);
+
+impl FromStr for Guid {
+    type Err = error::Error;
+    /// String is expected to look like "fe80:0000:0000:0000:e09d:7303:003f:3bf8"
+    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
+        let potential_guid = s.trim().to_string();
+        ensure!(
+            potential_guid.starts_with("fe80"),
+            error::InvalidGuidStringSnafu {
+                guid: potential_guid
+            }
+        );
+
+        let guid = Guid(s.trim().to_string());
+        Ok(guid)
+    }
+}
+
+impl Guid {
+    /// From the docs on usage around this GUID:
+    /// A U64 bit number queried from the CX device to allow FM to communicate with underlying NVSwitches.
+    /// We use this to trim down the second half of the value which is the required part for this functionality
+    ///
+    /// example "fe80:0000:0000:0000:e09d:7303:003f:3bf8" -> "0xe09d7303003f3bf8"
+    pub(crate) fn to_config_string(&self) -> String {
+        let full_string = self.0.to_string();
+        let parts: Vec<&str> = full_string.split(":").collect();
+        if parts.len() == 8 {
+            return format!("0x{}{}{}{}", parts[4], parts[5], parts[6], parts[7]);
+        }
+        self.0.to_string()
+    }
+}
+
+/// Capability Mask struct to help with FromStr and bit masking
+#[derive(Clone, Debug)]
+pub(crate) struct CapabilityMask(u32);
```

Comment:
You could also `impl Deref` for `CapabilityMask` and delegate everything to its inner `u32`.